### PR TITLE
Add Function to Preprocess Request Prompt & Address String Escape Symbol Issue

### LINKED ISSUE #278: GraphQL Syntax Error Due to String Escape Character in LangSmith Log History
**Describe the bug**
GraphQL Syntax Error spotted in LangSmith logs
```
Syntax Error GraphQL request (1:67) Invalid character escape sequence: \\\\\\\\\\\
```

**To Reproduce**
See one example log entry here
https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/421b4053-56ab-4704-a77d-33c374e4eac1?columnVisibilityModel=%7B%22outputs%22%3Afalse%2C%22feedback_stats%22%3Atrue%2C%22reference_example%22%3Afalse%2C%22first_token_time%22%3Afalse%2C%22tags%22%3Afalse%2C%22metadata%22%3Afalse%7D&timeModel=%7B%22duration%22%3A%227d%22%7D&tab=0&interval=1h&peek=a5dd3bee-698e-44e8-96bb-045124d5a073

1. Enter any prompt as long as it contains a SINGLE backslash by itself. Such as `\` or `123\123`
2. However, multiple backslashs together do not cause an error, such as `\\` or `text\\'`
3. See Error

**Expected behavior**
Should not error out

### PR DESCRIPTION
**Pull request by davidgxue**
### Description
- See https://github.com/astronomer/ask-astro/issues/278 for details of why the error occurs. This won't be explained again in this PR.
- After investigation, it seems like LangChain's implementation, specifically how Langchain calls Weaviate's hybrid retriever is not optimal.  Specifically, see this discussion thread from Weaviate https://forum.weaviate.io/t/hybrid-search-resulting-in-graphql-error/539. Technically, the safer way to build the query is using `.build()` instead of `.do()`. Using `build()` would have prevented symbols such as `\` in the original string to be properly transformed into GraphQL syntax compatible string. However, LangChain implemented this part of the code using `.do()` which causes issues with certain symbols that conflict with the GraphQL syntax. See here https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/retrievers/weaviate_hybrid_search.py#L152 on line 152. 
- As a result, we can only work around this issue by getting rid of the problem causing string escape symbol `\`. We can achieve this by parsing any single standalone `\` used to do string escape out of the original query. This shouldn't affect the original meaning of the prompt as the `\` gets auto added downstream anyway when querying the LLMs with string escape needed so that is not a concern.

### Technical Changes
- Add a `_preprocess_request` function that does input validation and transformation on the user prompt to workaround this issue in LangChain.

### Tests
1. When question exceeds character limit
image

2. When thread has more 6 questions/answers pairs, checked langsmith and saw only most recent 5 pairs 

3. Questions with more than 1 `\` is kept the same and questions with a single `\` the `\` is removed.
For example, the original github issue's error causing query is now being successfully executed.
Enter prompt: `KeyError: \'gcp\'` (note in Langsmith logs the prompt it auto adds another backslack to existing backslashs`)

image

closes #278

### CODE REVIEW

File: api/ask_astro/services/questions.py [hunk 1]
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
     )
 
 
+def _preprocess_request(request: AskAstroRequest) -> None:
+    if len(request.prompt) > 20000:
+        error_msg = "Question text is too long. Please try making a new thread and shortening your question."
+        request.response = error_msg
+        raise Exception(error_msg)
+    if len(request.prompt) == 0:
```
Comment (Lee-W) [inline api/ask_astro/services/questions.py:36]:
```suggestion
    if not request.prompt:
```

File: api/ask_astro/services/questions.py [hunk 2]
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
     )
 
 
+def _preprocess_request(request: AskAstroRequest) -> None:
+    if len(request.prompt) > 20000:
+        error_msg = "Question text is too long. Please try making a new thread and shortening your question."
+        request.response = error_msg
+        raise Exception(error_msg)
```
Comment (Lee-W) [inline api/ask_astro/services/questions.py:35]: I might suggestion adding a custom exception instead of using `Execption`

Comment (sunank200) [reply api/ask_astro/services/questions.py:35]: +1

File: api/ask_astro/services/questions.py [hunk 3]
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
...
+    if len(request.prompt) == 0:
+        error_msg = "Question text cannot be empty. Please try again with a different question."
+        request.response = error_msg
+        raise Exception(error_msg)
+    if len(request.messages) > 10:
+        request.messages = request.messages[-10:]
```
Comment (sunank200) [inline api/ask_astro/services/questions.py:41]: why -10?

**Comment by davidgxue** [comment]
Addressed your comment about custom error type in the new commit. I changed up the error propagation logic a little bit as well to make it cleaner. Thanks for the help reviewing @Lee-W  @sunank200

File: api/ask_astro/services/questions.py [hunk 4]
```diff
@@ -27,6 +28,20 @@ async def _update_firestore_request(request: AskAstroRequest) -> None:
...
 
+def _preprocess_request(request: AskAstroRequest) -> None:
+    if len(request.prompt) > 20000:
+        error_msg = "Question text is too long. Please try making a new thread and shortening your question."
+        request.response = error_msg
+        raise Exception(error_msg)
+    if not request.prompt:
+        error_msg = "Question text cannot be empty. Please try again with a different question."
+        request.response = error_msg
+        raise Exception(error_msg)
```
Comment (Lee-W) [inline api/ask_astro/services/questions.py:39]: Could we use a custom exception instead of using Execption here?

Comment (davidgxue) [reply api/ask_astro/services/questions.py]: Oops my bad I forgot to push my changes haha. I was like I thought I fixed this. Please feel free to re-review and resolve this comment if it's good!

File: api/ask_astro/services/questions.py (hunk 3 above)
Comment (sunank200) [reply api/ask_astro/services/questions.py:50]: Should this be a parameter instead so that it can be changed as part of env in deployment?

**Review by sunank200 — APPROVED** [review]

**Review by Lee-W — APPROVED** [review]

File: api/ask_astro/services/questions.py (hunk 3 above)
Comment (davidgxue) [reply api/ask_astro/services/questions.py:50]: Fixed!

**Pull request merged**